### PR TITLE
a better error message for disallowed trailing commas/additional parameters after rest elements in function params

### LINKED ISSUE #7460: Add a better error message for disallowed trailing commas after rest elements in function params
## Update

Let's make a nice "A trailing comma is not permitted after the rest element" error for all commas after function params. Currently we throw a nice error for arrow functions in this case, but apparently not for non-arrow functions.

## Original Issue below

This is a bug report.

### Input Code

```js
// error-input.js
function foo(
    first,
    ...rest
) {
}

// trailing-input.js
function foo(
    first,
    rest,
) {
}

// error-input.js
function foo(
    first,
    ...rest,
) {
}
```

### Babel/Babylon Configuration (.babelrc, package.json, cli command)

```
// package.json
{
  "dependencies": {
    "@babel/cli": "7.0.0-beta.40",
    "@babel/core": "7.0.0-beta.40"
  }
}
```

### Expected Behavior

```
> yarn -s babel rest-input.js
function foo(first, ...rest) {}

> yarn -s babel trailing-input.js
function foo(first, rest) {}

> yarn -s babel error-input.js
function foo(first, ...rest) {}
```

### Current Behavior

```
> yarn -s babel rest-input.js
function foo(first, ...rest) {}

> yarn -s babel trailing-input.js
function foo(first, rest) {}

> yarn -s babel error-input.js
{ SyntaxError: error-input.js: Unexpected token, expected ")" (3:11)

  1 | function foo(
  2 |     first,
> 3 |     ...rest,
    |            ^
  4 | ) {
  5 | }
  6 |
    at Parser.raise (C:\code\personal\bugs\babel-ts-traling-comma\node_modules\babylon\lib\index.js:772:15)
    at Parser.unexpected (C:\code\personal\bugs\babel-ts-traling-comma\node_modules\babylon\lib\index.js:2047:16)
    at Parser.expect (C:\code\personal\bugs\babel-ts-traling-comma\node_modules\babylon\lib\index.js:2035:28)
    at Parser.parseBindingList (C:\code\personal\bugs\babel-ts-traling-comma\node_modules\babylon\lib\index.js:2324:14)
    at Parser.parseFunctionParams (C:\code\personal\bugs\babel-ts-traling-comma\node_modules\babylon\lib\index.js:4560:24)
    at Parser.parseFunction (C:\code\personal\bugs\babel-ts-traling-comma\node_modules\babylon\lib\index.js:4548:10)
    at Parser.parseFunctionStatement (C:\code\personal\bugs\babel-ts-traling-comma\node_modules\babylon\lib\index.js:4206:17)
    at Parser.parseStatementContent (C:\code\personal\bugs\babel-ts-traling-comma\node_modules\babylon\lib\index.js:3923:21)
    at Parser.parseStatement (C:\code\personal\bugs\babel-ts-traling-comma\node_modules\babylon\lib\index.js:3899:17)
    at Parser.parseBlockOrModuleBlockBody (C:\code\personal\bugs\babel-ts-traling-comma\node_modules\babylon\lib\index.js:4426:23)
  pos: 38,
  loc: Position { line: 3, column: 11 },
  code: 'BABEL_PARSE_ERROR' }
error An unexpected error occurred: "Command failed.
Exit code: 1
Command: C:\\windows\\system32\\cmd.exe
Arguments: /d /s /c C:\\code\\personal\\bugs\\babel-ts-traling-comma\
ode_modules\\.bin\\babel error-input.js
Directory: C:\\code\\personal\\bugs\\babel-ts-traling-comma
Output:
".
info If you think this is a bug, please open a bug report with the information provided in "C:\\code\\personal\\bugs\\babel-ts-traling-comma\\yarn-error.log".
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

### Your Environment

| software         | version(s)
| ---------------- | -------
| Babel            |  7.0.0-beta.40
| node             | v9.6.1
| npm              | yarn 1.5.1
| Operating System | Win 10

### PR DESCRIPTION
**Pull request by morozRed**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Fixes #7460  
| Patch: Bug Fix?          | n
| Major: Breaking Change?  | n
| Minor: New Feature?      | n
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

This PR changes error message when we had other parameters after rest element.

**Previous behaviour:**

Test:
```
(
    first,
    ...second,
    third
) => {};
```
Result
```
Unexpected token, expected ")"
```
**New behaviour:**

Test:
```
(
    first,
    ...second,
    third
) => {};
```

Result:
```
Rest parameter must be last formal parameter
```

### CODE REVIEW

File: packages/babel-parser/src/parser/expression.js [hunk 1]
```diff
@@ -1120,6 +1120,17 @@ export default class ExpressionParser extends LValParser {
...
         }
 
+        if (
+          this.match(tt.comma) &&
+          (this.lookahead().type === tt.name ||
+            this.lookahead().type === tt.ellipsis)
+        ) {
+          this.raise(
+            this.state.start,
+            "Rest parameter must be last formal parameter",
```
Comment (xtuc) [inline packages/babel-parser/src/parser/expression.js]: Thanks for your PR, I looks good.

At first the error mesage looked odd to me, but it's actually V8's one :+1:

Comment (morozRed) [reply packages/babel-parser/src/parser/expression.js]: @xtuc thanks! I'm glad that my first PR was not bad 😃 Yeah, I decided to get this message from V8.

File: packages/babel-parser/src/parser/expression.js [hunk 2]
```diff
@@ -1120,6 +1120,17 @@ export default class ExpressionParser extends LValParser {
           );
         }
 
+        if (
+          this.match(tt.comma) &&
+          (this.lookahead().type === tt.name ||
+            this.lookahead().type === tt.ellipsis)
```
Comment (nicolo-ribaudo) [inline packages/babel-parser/src/parser/expression.js]: We don't need the lookahead, since also `(a, ...b,) => {}` is invalid.

Comment (morozRed) [reply packages/babel-parser/src/parser/expression.js]: @nicolo-ribaudo right, so basically we can do just:
```
if (this.match(tt.comma)) {
          this.raise(
            this.state.start,
            "A trailing comma is not permitted after the rest element",
          );
        }
```

Comment (morozRed) [reply packages/babel-parser/src/parser/expression.js]: @nicolo-ribaudo so you want to leave only this type of error? 
`"A trailing comma is not permitted after the rest element"`

Comment (existentialism) [reply packages/babel-parser/src/parser/expression.js]: We can combine the two conditionals and do the lookahead once?

Comment (nicolo-ribaudo) [reply packages/babel-parser/src/parser/expression.js]: I agree with @existentialism. Btw, in the lookahead we should check for `}` and `]`.

Comment (morozRed) [reply packages/babel-parser/src/parser/expression.js]: @existentialism @nicolo-ribaudo so maybe it's even better to add switch there

Comment (nicolo-ribaudo) [reply packages/babel-parser/src/parser/expression.js]: Well, there are only two cases:
1. Trailing comma
2. Another parameter

Comment (morozRed) [reply packages/babel-parser/src/parser/expression.js]: @nicolo-ribaudo what about `[` and `{`?

Comment (morozRed) [reply packages/babel-parser/src/parser/expression.js]: @nicolo-ribaudo @existentialism So something like this

```
if (this.match(tt.comma)) {
    const nextTokenType = this.lookahead().type;
    const errorMessage =
    nextTokenType === tt.parenR
        ? "A trailing comma is not permitted after the rest element"
        : "Rest parameter must be last formal parameter";
    this.raise(this.state.start, errorMessage);
}
```

Comment (existentialism) [reply packages/babel-parser/src/parser/expression.js]: 👍

Comment (nicolo-ribaudo) [reply packages/babel-parser/src/parser/expression.js]: Yes :+1:

**Comment by nicolo-ribaudo** [comment]
Could you add a test for `(...rest,) => {}`?

**Comment by morozRed** [comment]
@nicolo-ribaudo yeah, sure

**Comment by nicolo-ribaudo** [comment]
Thank you!

**Pull request merged**